### PR TITLE
Use the certificate role to create the cert and the key

### PR DESCRIPTION
**Pull request by nhosoi**
- Introduce a variable cockpit_certificates to set the certificate_requests.
- Add a test test script tests/tests_certificate2.yml

I borrowed (again :) the mssql README in https://github.com/linux-system-roles/mssql/pull/125 written by @spetrosi. Thanks, Sergei!

@spetrosi, @martinpitt , this is another piece of the effort to use other roles in the cockpit role. If you could review it, I'd greatly appreciate it.

### CODE REVIEW

**Comment by nhosoi** [comment]
[citest]

File: README.md [hunk 1]
```diff
@@ -170,6 +172,35 @@ You can also use `ca: self-sign` or `ca: local` depending on your certmonger usa
 
 Note that this does *not* work on RHEL/CentOS 7.
 
+#### Generate a new certificate in the role
+
+You can also use the `certificate` role inside the `cockpit` role to create
+certificates by providing `cockpit_certificates`.
+
+Use this variable to generate certificate and private key for TLS encryption
```
Comment (spetrosi) [inline README.md:180]:
```suggestion
Use the `cockpit_certificates` variable to generate certificate and private key for TLS encryption
```

Comment (nhosoi) [reply README.md]: Thanks, @spetrosi!

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosoi** [comment]
[citest pending]

**Comment by nhosoi** [comment]
@spetrosi, @martinpitt, all the CI tests have passed. Could you please review one more time? Thanks!

**Review by spetrosi — DISMISSED** [review]
lgtm

**Comment by nhosoi** [comment]
Thanks a lot for your reviews, @spetrosi.
@martinpitt, could you please take another look? If it looks ok to you, may I ask you to merge this pr? Thanks, in advance.

File: README.md [hunk 2]
```diff
@@ -170,6 +172,35 @@ You can also use `ca: self-sign` or `ca: local` depending on your certmonger usa
 
 Note that this does *not* work on RHEL/CentOS 7.
 
+#### Generate a new certificate in the role
```
Comment (martinpitt) [inline README.md:175]: I find this a bit nicer, and our role retains more control over what actually happens. So :+1: from me for moving to that approach.

But please let's actually *move* -- I find it confusing to offer "you can do it like this or that" in the README. Which one is better? (from the admin's POV). IMHO we should recommend just one approach.

The `setype:` hack is indeed obsolete. RHEL 8.6 has Cockpit 264, and CentOS 8 has something much newer. So we don't need this any more, neither with the "builtin" role.

File: README.md [hunk 3]
```diff
@@ -170,6 +172,35 @@ You can also use `ca: self-sign` or `ca: local` depending on your certmonger usa
...
+
+You can also use the `certificate` role inside the `cockpit` role to create
+certificates by providing `cockpit_certificates`.
```
Comment (martinpitt) [inline README.md:178]: This is redundant with the paragraph below, and "providing" is confusing. If we drop the current approach from the section above, then this paragraph can just be dropped.

File: README.md [hunk 4]
```diff
@@ -170,6 +172,35 @@ You can also use `ca: self-sign` or `ca: local` depending on your certmonger usa
...
+
+Use the `cockpit_certificates` variable to generate certificate and private key
+for TLS encryption using the `fedora.linux_system_roles.certificate`.
+
```
Comment (martinpitt) [inline README.md:182]: Several grammar fixes:

> Set the `cockpit_certificates` variable to generate a certificate and private key for TLS using `fedora.linux_system_roles.certificate`.

File: README.md [hunk 5]
```diff
@@ -170,6 +172,35 @@ You can also use `ca: self-sign` or `ca: local` depending on your certmonger usa
...
+The value of `cockpit_certificates` is set to the variable `certificate_requests`
+defined in the `certificate` role.
+For more information, see the `certificate_requests` section in the `certificate`
+role documentation.
```
Comment (martinpitt) [inline README.md:186]: > `cockpit_certificates` is passed on to the `certificate_requests` variable of the certificate role. For more information, see the `certificate_requests` role documentation section.

The latter should also be linked, if at all possible.

File: README.md [hunk 6]
```diff
@@ -170,6 +172,35 @@ You can also use `ca: self-sign` or `ca: local` depending on your certmonger usa
...
+
+When you set `cockpit_certificates`, you must not set `cockpit_private_key` and
+`cockpit_cert` variables.
+
+This example installs the Cockpit with the Cockpit web server certificate.
```
Comment (martinpitt) [inline README.md:191]: > This example installs Cockpit with an IdM-issued web server certificate.

File: tests/tests_certificate2.yml [hunk 7]
```diff
@@ -0,0 +1,62 @@
+---
```
Comment (martinpitt) [inline tests/tests_certificate2.yml:1]: If we drop the old approach, you can update the existing tests_certificate.yml, and avoid the "huh, what's 2" duplication.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thank you! I'd like to see just *one* approach, not two. I also have some grammar/style fixes for the documentation.

File: README.md (hunk 2 above)
Comment (nhosoi) [reply README.md:175]: > I find this a bit nicer, and our role retains more control over what actually happens. So +1 from me for moving to that approach.
> 
> But please let's actually _move_ -- I find it confusing to offer "you can do it like this or that" in the README. Which one is better? (from the admin's POV). IMHO we should recommend just one approach.
> 
> The `setype:` hack is indeed obsolete. RHEL 8.6 has Cockpit 264, and CentOS 8 has something much newer. So we don't need this any more, neither with the "builtin" role.

Hmmm, a good point... Maybe, we can say something like this?
```
### Generate a new certificate

Now you can do it in the cockpit role by providing the cockpit_certificates variable as follows
(in the nicer English, of course ;)
    - name: Install cockpit with Cockpit web server certificate
      include_role:
        name: linux-system-roles.cockpit
      vars:
        cockpit_certificates:
          - name: monger-cockpit
            dns: ['localhost', 'www.example.com']
            ca: ipa
            group: cockpit-ws

If you already have a playbook using the certificate role and the cockpit role separately,
that's supported, too.
    - name: Generate Cockpit web server certificate
      include_role:
        name: linux-system-roles.certificate
      vars:
        certificate_requests:
          - name: /etc/cockpit/ws-certs.d/monger-cockpit
            dns: ['localhost', 'www.example.com']
            ca: ipa
            group: cockpit-ws
```

File: tests/tests_certificate2.yml (hunk 7 above)
Comment (nhosoi) [reply tests/tests_certificate2.yml:1]: It'd be nice to keep the original tests_certificate.yml, IMHO. I can rename it to, like, tests_certificate_internal.yml or something similar. (maybe, I can rename the original tests_certificate.yml to tests_certificate_external.yml, as well. :)

File: README.md (hunk 2 above)
Comment (martinpitt) [reply README.md:175]: Agreed -- documenting this new way as the primary method, and just giving this a fallback is fine.

File: tests/tests_certificate2.yml (hunk 7 above)
Comment (martinpitt) [reply tests/tests_certificate2.yml:1]: Yes, with a renaming it's clearer what they do.

**Comment by nhosoi** [comment]
[citest]

**Review by martinpitt — DISMISSED** [review]
Thank you!

**Comment by nhosoi** [comment]
[citest pending]

**Comment by nhosoi** [comment]
Thank you for your reviews and suggestions, @martinpitt and @spetrosi.
All the CI tests have passed.
If you have the privilege and merge this pr, I'd appreciate it. Thanks!

**Comment by nhosoi** [comment]
@richm, I'm adding your note to the cockpit README, as well.
Do you think my [addition](https://github.com/linux-system-roles/cockpit/pull/78/commits/78b071fc42dd485427bf902b8e00ecb464375ce2) is appropriate?

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosoi** [comment]
[citest pending]

**Comment by nhosoi** [comment]
I believe all the CI tests have passed.
This is a leftover from the old attempt...
```
RHEL-8.8.0-20221020.2/ansible-2.9/(citool) Pending — Test started
```
@richm, @martinpitt, could you please review one more time (if needed) and merge this pr? Thanks, in advance.

**Review by richm — DISMISSED** [review]

File: tasks/main.yml [hunk 8]
```diff
@@ -79,6 +79,23 @@
   when: cockpit_config is defined
   notify: restart cockpit
 
+- name: Create certificates
+  when:
+    - cockpit_certificates | length > 0
+    - ansible_facts['os_family'] == 'RedHat'
```
Comment (richm) [inline tasks/main.yml:85]: The node in the readme says `NOTE: This does *not* work on RHEL/CentOS 7.` - do we need to add a condition for that e.g.
```yaml
    - ansible_facts['distribution_version'] is version('>=', '8')
```
?

Comment (nhosoi) [reply tasks/main.yml:85]: Ahh, sorry. The note is not needed. The issue was if we try to generate the new cert directly in the target location `/etc/cockpit/ws-certs.d` like this:
```
    cockpit_certificates:
      - name: /etc/cockpit/ws-certs.d/cockpit_cert
        <<snip>>
```
it fails due to [this bug](https://github.com/linux-system-roles/certificate/pull/97) on RHEL/CentOS-7 with an invalid mode.

In this PR, it generates the private key and cert in the standard location and copies them to the directories Cockpit expects. So, it is free from the issue. And there's no problem running on RHEL/CentOS-7. (CI successfully passed on the platforms.) I'm removing the note. Thanks for finding it out, @richm!

Comment (richm) [reply tasks/main.yml:85]: > Do we want to describe in a little more detailed manner something like this?

Yes - and - do we have a bz or some other issue tracker for this issue?

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosoi** [comment]
> > Do we want to describe in a little more detailed manner something like this?
>
> Yes - and - do we have a bz or some other issue tracker for this issue? Or maybe it is a "feature" of the cockpit/certmonger?

These 2 issues were filed by @martinpitt against RHEL/CentOS-7 last year, which are still open.
https://github.com/linux-system-roles/certificate/issues/98
https://github.com/linux-system-roles/certificate/issues/99

And as you suggested, I've added a condition - if the distribution is `RedHat` and the version is `7`, and the ca is `self-sign`, it fails with an error `Creating a self-signed certificate is not supported on RHEL-7` or `CentOS-7`.

File: tasks/main.yml [hunk 9]
```diff
@@ -79,6 +79,32 @@
...
+      fail:
+        msg: >-
+          Creating a self-signed certificate is not supported on
+          {{ ansible_facts['distribution'] }}-{{ ansible_facts['distribution_version'] }}
+
+    - name: Create certificates using the certificate role
+      include_role:
+        name: fedora.linux_system_roles.certificate
+      vars:
+        __cert_name: "{{ cockpit_certificates.0.name  | basename }}"
```
Comment (richm) [inline tasks/main.yml:100]: where is `__cert_name` used?

Comment (nhosoi) [reply tasks/main.yml]: Thank you, @richm. Removed it.

**Review by richm — APPROVED** [review]

**Comment by nhosoi** [comment]
[citest]

**Pull request merged**